include/zephyr/bindesc.h

```diff
@@ -0,0 +1,398 @@
+#endif /* IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_STRING) */
+
+#if IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_MAJOR)
+extern const struct bindesc_entry BINDESC_NAME(kernel_version_major);
+#endif /* IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_MAJOR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_MINOR)
+extern const struct bindesc_entry BINDESC_NAME(kernel_version_minor);
+#endif /* IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_MINOR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_PATCHLEVEL)
+extern const struct bindesc_entry BINDESC_NAME(kernel_version_patchlevel);
+#endif /* IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_PATCHLEVEL) */
+
+#if IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_NUMBER)
+extern const struct bindesc_entry BINDESC_NAME(kernel_version_number);
+#endif /* IS_ENABLED(CONFIG_BINDESC_KERNEL_VERSION_NUMBER) */
+
+#if IS_ENABLED(CONFIG_BINDESC_APP_VERSION_STRING)
+extern const struct bindesc_entry BINDESC_NAME(app_version_string);
+#endif /* IS_ENABLED(CONFIG_BINDESC_APP_VERSION_STRING) */
+
+#if IS_ENABLED(CONFIG_BINDESC_APP_VERSION_MAJOR)
+extern const struct bindesc_entry BINDESC_NAME(app_version_major);
+#endif /* IS_ENABLED(CONFIG_BINDESC_APP_VERSION_MAJOR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_APP_VERSION_MINOR)
+extern const struct bindesc_entry BINDESC_NAME(app_version_minor);
+#endif /* IS_ENABLED(CONFIG_BINDESC_APP_VERSION_MINOR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_APP_VERSION_PATCHLEVEL)
+extern const struct bindesc_entry BINDESC_NAME(app_version_patchlevel);
+#endif /* IS_ENABLED(CONFIG_BINDESC_APP_VERSION_PATCHLEVEL) */
+
+#if IS_ENABLED(CONFIG_BINDESC_APP_VERSION_NUMBER)
+extern const struct bindesc_entry BINDESC_NAME(app_version_number);
+#endif /* IS_ENABLED(CONFIG_BINDESC_APP_VERSION_NUMBER) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_YEAR)
+extern const struct bindesc_entry BINDESC_NAME(build_time_year);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_YEAR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MONTH)
+extern const struct bindesc_entry BINDESC_NAME(build_time_month);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MONTH) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_DAY)
+extern const struct bindesc_entry BINDESC_NAME(build_time_day);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_DAY) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_HOUR)
+extern const struct bindesc_entry BINDESC_NAME(build_time_hour);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_HOUR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MINUTE)
+extern const struct bindesc_entry BINDESC_NAME(build_time_minute);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MINUTE) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_SECOND)
+extern const struct bindesc_entry BINDESC_NAME(build_time_second);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_SECOND) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_UNIX)
+extern const struct bindesc_entry BINDESC_NAME(build_time_unix);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_UNIX) */
```

Comment:
IMHO this is overkill. It should be enough to just have epoch string, which will be far more efficient than doing these separately.

Reply:
I added them in case someone will ever find it useful. I agree that the individual year/month/day etc. is a very inefficient way to store the build time, but you don't pay for it if you don't use it so I see no harm in having them.